Support access log from sidecar and per-endpoint redirect stats

rlenglet opened the pull request:
Use the policy name instead of listener ID in Envoy access logs.

Remove duplicate `accesslog.pb.go` file. Move files related to access log protobufs into the right package (`pkg/envoy/cilium`). Remove `listener_id` from filter configuration. Use network policy
resource name instead.

Extract network policy name and observation point from each log record, instead of getting it from an association with a redirect / listener. Instead, associate access log producers to network policies.

Add the observation point (ingress / egress) in every Envoy log record, and use that to create an access log record.

Remove all logging logic from `Redirect`.    
Remove all stats logic from `Redirect`, to decouple stats from redirects to local proxies.

Split the `LogRecordProducer` interface into the existing `LogRecordNotifier` interface and a new `EndpointInfoRegistry` interface. Group identity lookup by ID and by IP under the same `EndpointInfoRegistry` interface.
    
Split files for readability:

- `logger.go: move endpoint info interface into `epinfo.go`
- `proxy.go`: move `Redirect` into `redirect.go`, `getMagicMark` into `mark.go`

Move redirect statistics from proxy status to endpoint API. Remove statistics from `ProxyRedirectStatus`. Add security identity and labels SHA256 to `ProxyRedirectStatus`, for consistency.
    
Move `ProxyRedirectStatistics` into `Endpoint` (i.e. `cilium endpoint get`) and redefine it as a subset of `ProxyRedirectStatus` plus statistics.

The output of `cilium endpoint get` now returns stats like:

```
        "proxy-statistics": [
          {
            "location": "ingress",
            "port": 80,
            "protocol": "http",
            "statistics": {
              "requests": {
                "denied": 12,
                "forwarded": 15,
                "received": 27
              },
              "responses": {
                "forwarded": 15,
                "received": 15
              }
            }
          }
        ],
```

Rename the `NetworkPolicyEndpoint` interface into `EndpointUpdater`. Move `EndpointUpdater` from `pkg/envoy` into `pkg/proxy/logger` to make usable from the Kafka proxy.

Update stats from Kafka proxy after logging into access log.
    
Fix access logging to only log and account once when a request is denied and a response can't be created.

Add endpoint identities, IP addresses, and ports in Kafka response access logs. This is also required to be able to derive endpoint stats.

Look up original destination address and port of a Kafka request connection only once after accepting the connection, instead of for every request.

Simplify the implementation of the Kafka proxy's `handleRequests` function as it was always passed a `nil` log record.

Remove the `redirects` field from proxy status in `cilium status`. Move the `allocated-proxy-port` field into `proxy-statistics` in `cilium endpoint get`.

Sort the elements in `proxy-statistics`.

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
envoy: Use policy name instead of listener ID in access logs

Remove duplicate accesslog.pb.go. Move files related to access log
protobufs into the right package: pkg/envoy/cilium.
Remove listener_id from filter configuration. Use network policy
resource name instead.

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
envoy: Extract network policy name and obs point from log records

Associate access log producers to network policies.
Remove all logging logic from Redirect.

Add the observation point (ingress / egress) in every Envoy log
record, and use that to create an access log record.

Remove all stats logic from Redirect, to decouple stats from
redirects to local proxies.

Split the LogRecordProducer interface into the existing
LogRecordNotifier interface and a new EndpointInfoRegistry
interface. Group identity lookup by ID and by IP under the same
EndpointInfoRegistry interface.

Split files for readability:
logger.go: move enpoint info interface into epinfo.go
proxy.go: move Redirect into redirect.go, getMagicMark into mark.go

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
api: Move redirect statistics from proxy status to endpoint

Remove statistics from ProxyRedirectStatus.
Add security identity and labels SHA256 to ProxyRedirectStatus.

Move ProxyRedirectStatistics into Endpoint (cilium endpoint get)
and redefine it as a subset of ProxyRedirectStatus plus statistics.

The output of cilium endpoint get now contains stats like:

    "proxy-statistics": [
      {
        "location": "ingress",
        "port": 80,
        "protocol": "http",
        "statistics": {
          "requests": {
            "denied": 12,
            "forwarded": 15,
            "received": 27
          },
          "responses": {
            "forwarded": 15,
            "received": 15
          }
        }
      }
    ],

Signed-off-by: Romain Lenglet <romain@covalent.io>

tgraf commented on pkg/proxy/logger/logger.go:
```diff
@@ -62,80 +62,95 @@ const (
-// fillEndpointInfo tries to resolve the IP address and fills the EndpointInfo
-// fields with either ReservedIdentityHost or ReservedIdentityWorld
-func fillEndpointInfo(producer LogRecordProducer, info *accesslog.EndpointInfo, ip net.IP) {
+// fillEndpointInfo tries to resolve the IP address and return the EndpointInfo
```
The adjusted comment doesn't make sense

rlenglet replied:
Reverted. This was leftover from another change I tried.

tgraf commented on pkg/proxy/logger/logger.go:
```diff
@@ -217,11 +230,11 @@ type AddressingInfo struct {
-			lr.DestinationEndpoint = lr.producer.LocalEndpointInfo()
-		case accesslog.Egress:
-			lr.SourceEndpoint = lr.producer.LocalEndpointInfo()
+		ingress := lr.ObservationPoint == accesslog.Ingress
```
Not sure this style change is really necessary

rlenglet replied:
It's an optimization. 1 string comparison instead of 4.
That's not essential. I reverted this.

tgraf reviewed (dismissed):
Will review the rest later

ianvernon reviewed (approved):
API and endpoint changes look good. Having such visibility per endpoint is an awesome addition :D

tgraf commented on api/v1/openapi.yaml:
```diff
@@ -824,8 +824,13 @@ definitions:
+        description: Statistics of the proxy redirects configured for this endpoint
+        type: array
+        items:
+          "$ref": "#/definitions/ProxyRedirectStatistics"
```
This split seems weird, why not just encode the entire proxy redirect status in the endpoint? Splitting this means a user has to manually correlate the statistics via the endpoint and rules and proxy port via the global status.

rlenglet replied:
Because we won't have that information from sidecar proxies. The only information that we won't have is the proxy redirect port. I don't think that's essential.

rlenglet replied:
One advantage of aggregating stats at endpoint level is that we'll have stats that will span the whole lifetime of the endpoint, not just the lifetime of redirects. I think that's more useful. For instance, we can still see the total count of denied requests even after policy changes that result in redirect changes.

rlenglet replied:
And we can't correlate stats from sidecar proxies with redirects, because we don't even configure redirects in this case. That's the whole point of this PR.

tgraf replied:
I agree that having the stats in the endpoint is neat. I also understand that there is no conventional proxy port, although it would be great to see the port number on which Envoy is listening. My main concern left is that in order to see the proxy rules, the user has to correlate the entry in the status.

Users will also get used to seeing the proxy status in either place and it will just be empty for the sidecar model. A user will not want to understand what a redirect and why it is missing for the sidecar case. Can't we move everything to the endpoint and just not call it a redirect?

rlenglet replied:
We can't get the redirect port from sidecars we don't manage, e.g. Istio sidecars. And even if we did it wouldn't be useful information. That port is 15000 by default for every redirect across the whole cluster.

I could add the redirect port into the Endpoint's `proxy-statistics` if there's a redirect currently active for the port *and* if the redirect port is known:
```
 "proxy-statistics": [
          {
            "location": "ingress",
            "port": 80,
            "protocol": "http",
            "redirect-port": 12345,
            "statistics": {
              ...
            }
          }
        ],
```

And if there's currently no redirect for the port, or the redirect port is not known, there will be no `"redirect-port"` field:
```
 "proxy-statistics": [
          {
            "location": "ingress",
            "port": 80,
            "protocol": "http",
            "statistics": {
              ...
            }
          }
        ],
```

And then I can remove the redirects from `cilium status` altogether.

The only drawback is that it will be hard to know at a glance what redirects are currently installed into the Envoy and Kafka proxies managed by Cilium (not in sidecars).

@tgraf, would you be OK with that API?

rlenglet replied:
And regarding the rules enforced in the proxy, now it’s even neater because with NPDS we now push the endpoints’ full policies to the Envoy proxies. And an endpoint’s policy is right in the same object returned by `cilium endpoint get`.

When we move the Kafka proxy to use xDS / NPDS for configuration, that will also be true for Kafka. We will NOT configure rules specifically for each redirect.

tgraf replied:
> That port is 15000 by default for every redirect across the whole cluster.

It may be clear to you, it will not be clear to users and it will be useful information to the user as there will be traffic inside of the pod to and from that port. As long as we can add it later on, I'm fine.

> And regarding the rules enforced in the proxy, now it’s even neater because with NPDS we now push the endpoints’ full policies to the Envoy proxies. And an endpoint’s policy is right in the same object returned by cilium endpoint get.

That's great, given we are just before 1.0, we absolutely don't want to expose something in the overall status and then remove it. How difficult is it to just expose the kafka rules like this as well? I really dislike exposing any information in different places depending on whether the proxy model is sidecar or not.

rlenglet replied:
We'll have to move the Kafka proxy to use NPDS when we move it into a separate process. When that's done, the policy that is returned in `cilium endpoint get` will be the one that is pushed to the Kafka proxy.

I'll remove `redirects` altogether from the proxy status then.

tgraf reviewed (changes requested):
What is the state of Kafka with this current PR? Does this break Kafka?

rlenglet committed:
proxy: Update endpoint stats from Kafka proxy

Rename NetworkPolicyEndpoint interface into EndpointUpdater.
Move EndpointUpdater from pkg/envoy into pkg/proxy/logger to make
usable from the Kafka proxy.

Update stats from Kafka proxy after logging into access log.

Fix access logging to only log and account once when a request is
denied and a response can't be created.

Add endpoint identities, IP addresses, and ports in Kafka response
access logs. This is also required to be able to derive endpoint
stats.

Look up original destination address and port of a Kafka request
connection only once after accepting the connection, instead of
for every request.

Simplify the implementation of handleRequests as it was always
passed a nil log record.

Signed-off-by: Romain Lenglet <romain@covalent.io>

jrajahalme reviewed (approved):
LGTM

manalibhutiyani commented on pkg/proxy/kafka.go:
```diff
@@ -322,20 +346,17 @@ func handleRequests(done <-chan struct{}, pair *connectionPair, c *proxyConnecti
 		}
 
 		if err != nil {
-			if record != nil {
```
Why was this removed? Shouldn't we be putting this in access log?

rlenglet replied:
record was always nil. This was dead code.

rlenglet committed:
api: Remove redirects from proxy status

Remove the redirects field from proxy status in "cilium status".
Move allocated-proxy-port field into proxy-statistics in
"cilium endpoint get".

Sort the elements of proxy-statistics.

Signed-off-by: Romain Lenglet <romain@covalent.io>
